### PR TITLE
Graphql/user

### LINKED ISSUE #1: User Query - User Profile Information
As a frontend developer querying the CodeHerd API, I am able to send a a user id and receive a JSON payload with each or any (as keys) of the following that I request:
- id
- name
- email
- slack handle
- github handle
- an array of their preferred working styles
- their cohort number
- their mod number?
- github_access_token
- pronouns

### PR DESCRIPTION
**Pull request by zachjamesgreen**
## Checklist
- [x] Tests written where required
- [x] SimpleCov at 56.04%
- [ ] Squash & Rebase
- [x] All tests passing
- [ ] Prettier has been run
- [x] Reek has been run
- [x] Circle CI builds are passing

## Changes
Can query a user, all users, and edit a user
added graphql schema file
added some reek configs

closes #1

### CODE REVIEW

File: spec/rails_helper.rb [hunk 1]
```diff
@@ -75,4 +75,5 @@
   end
 end
 
-OmniAuth.config.test_mode = true
\ No newline at end of file
+OmniAuth.config.test_mode = true
+OmniAuth.config.silence_get_warning = true
```
Comment (twarbelow) [inline spec/rails_helper.rb:79]: what does this do?

File: spec/request/get users_spec.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
+require 'rails_helper'
```
Comment (twarbelow) [inline spec/request/get users_spec.rb]: will we ever need to get all users?

File: schema.graphql [hunk 3]
```diff
@@ -0,0 +1,60 @@
+"""
```
Comment (twarbelow) [inline schema.graphql:1]: are the contents of this file autogenerated?

File: app/graphql/types/query_type.rb [hunk 4]
```diff
@@ -8,10 +8,26 @@ class QueryType < Types::BaseObject
...
+
+    field :users, [Types::UserType], null: false,
+      description: "return all users"
+
+    field :user, Types::UserType, null: false do
+      description "return a user"
+      argument :id, ID, required: true
+    end
+
+    def users
```
Comment (twarbelow) [inline app/graphql/types/query_type.rb:25]: not sure when we would need this...

File: app/graphql/types/queries/user_queries.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
+module Queries
+  class User < Queries::BaseQuery
+    def allUsers
```
Comment (twarbelow) [inline app/graphql/types/queries/user_queries.rb:3]: could you explain the difference between putting this in types/queries/user_queries.rb and having a super similar one in types/query_type?

File: Rakefile [hunk 6]
```diff
@@ -1,6 +1,11 @@
 # Add your own tasks in files placed in lib/tasks ending in .rake,
 # for example lib/tasks/capistrano.rake, and they will automatically be available to Rake.
+require 'graphql/rake_task'
 
 require_relative 'config/application'
 
 Rails.application.load_tasks
+
```
Comment (twarbelow) [inline Rakefile:8]: what does this do?

File: app/controllers/sessions_controller.rb [hunk 7]
```diff
@@ -2,13 +2,19 @@ class SessionsController < ApplicationController
...
   def new
     # This route is catched by the Omniauth Middleware and is invisible to rake routes
+    # TODO: check if user is already in database
     redirect_to '/auth/github'
   end
 
   def create
-    User.find_or_create_from_auth_hash(auth_hash)
+    user = User.find_or_create_from_auth_hash(auth_hash)
+    str = [user.id.to_s, user.name, user.email].to_s
```
Comment (twarbelow) [inline app/controllers/sessions_controller.rb:14]: do you need the to_s  on user.id as well as on the array?

File: app/controllers/sessions_controller.rb [hunk 8]
```diff
@@ -2,13 +2,19 @@ class SessionsController < ApplicationController
...
+    encoded_token = Base64.encode64(str)
```
Comment (twarbelow) [inline app/controllers/sessions_controller.rb:15]: have you talked to frontend about this? just so they know what to do once it gets back to them?

File: spec/rails_helper.rb (hunk 1 above)
Comment (zachjamesgreen) [reply spec/rails_helper.rb:79]: its a way of mocking out OmniAuth so we can test user creation without making a network call
the second line silence the error for allowing get requests with OAuth. needed to work with the github OAuth provider

File: app/graphql/types/query_type.rb [hunk 9]
```diff
@@ -8,10 +8,26 @@ class QueryType < Types::BaseObject
...
-      description: "An example field added by the generator"
-    def test_field
-      "Hello World!"
+    # field :test_field, String, null: false,
+    #   description: "An example field added by the generator"
+    # def test_field
+    #   "Hello World!"
+    # end
+
+    field :users, [Types::UserType], null: false,
```
Comment (twarbelow) [inline app/graphql/types/query_type.rb:17]: will we use this?

**Review by twarbelow — CHANGES_REQUESTED** [review]

File: spec/request/get users_spec.rb (hunk 2 above)
Comment (zachjamesgreen) [reply spec/request/get users_spec.rb]: no had it in there to get the hang of Graphql can delete on next PR

File: schema.graphql (hunk 3 above)
Comment (zachjamesgreen) [reply schema.graphql:1]: Yes they are using the task `rails graphql:schema:dump`

File: app/graphql/types/query_type.rb (hunk 4 above)
Comment (zachjamesgreen) [reply app/graphql/types/query_type.rb:25]: no had it in there to get the hang of Graphql can delete on next PR

File: app/graphql/types/queries/user_queries.rb (hunk 5 above)
Comment (zachjamesgreen) [reply app/graphql/types/queries/user_queries.rb:3]: This is an on going experiment to put queries that are related in their own file for better organization. right now all queries live in query_type file which will cause merge conflicts

File: Rakefile (hunk 6 above)
Comment (zachjamesgreen) [reply Rakefile:8]: This is the rake task that created the schema.graphql file and the schema.json file

File: app/controllers/sessions_controller.rb (hunk 7 above)
Comment (zachjamesgreen) [reply app/controllers/sessions_controller.rb:14]: yes to make the id which is an integer a string when parsed and to make the array an string

File: app/controllers/sessions_controller.rb (hunk 8 above)
Comment (zachjamesgreen) [reply app/controllers/sessions_controller.rb:15]: not yet will talk at standup tomorrow morn. also mentioned in late night standup

File: app/graphql/types/query_type.rb (hunk 9 above)
Comment (zachjamesgreen) [reply app/graphql/types/query_type.rb:17]: no had it in there to get the hang of Graphql can delete on next PR

**Review by twarbelow — APPROVED** [review]
thanks for the insight Zach

**Pull request merged**